src/utils/dnr-converter.js

```diff
@@ -9,131 +9,88 @@
 
-      const resolve = requestResolvers.get(requestId);
-
-      rules =
-        __PLATFORM__ === 'safari'
-          ? rules
-              .map((r) => {
-                try {
-                  return convert(r);
-                } catch (e) {
-                  errors.push(e);
-                }
-              })
-              .filter(Boolean)
-          : rules;
-
-      resolve({
-        rules,
-        errors: errors.map((e) => `DNR: ${e.message}`),
+      result = await chrome.runtime.sendMessage({
+        action: 'dnr-converter:convert',
+        filters,
       });
 
-      requestResolvers.delete(requestId);
-    });
-
-    const iframe = document.createElement('iframe');
-    iframe.setAttribute('src', 'https://ghostery.github.io/urlfilter2dnr/');
-    iframe.setAttribute('style', 'display: none;');
-
-    documentConverter = new Promise((resolve, reject) => {
-      iframe.addEventListener('load', () => resolve(iframe));
-      iframe.addEventListener('error', reject);
+      closeOffscreenDocument();
+    } else if (__PLATFORM__ === 'safari') {
+      const { convertWithAdguard } = await import('@ghostery/urlfilter2dnr');
+      result = await convertWithAdguard(filters);
+
+      result.rules = result.rules.reduce((acc, r) => {
+        try {
+          acc.push(convertToSafariFormat(r));
+        } catch (e) {
+          result.errors.push(e);
+        }
+        return acc;
+      }, []);
+      result.errors = result.errors.map((e) => `DNR: ${e.message}`);
+    } else {
+      throw new Error('Unsupported platform for DNR conversion');
+    }
+  } catch (e) {
+    return { errors: [e.message], rules: [] };
+  }
 
-      document.head.appendChild(iframe);
-    });
+  for (const [index, rule] of result.rules.entries()) {
+    if (rule.condition.regexFilter) {
+      const { isSupported, reason } =
+        await chrome.declarativeNetRequest.isRegexSupported({
+          regex: rule.condition.regexFilter,
+        });
 
-    return documentConverter;
+      if (!isSupported) {
+        result.errors.push(
+          `Could not apply a custom filter as "${rule.condition.regexFilter}" is a not supported regexp due to: ${reason}`,
+        );
+        result.rules.splice(index, 1);
+      }
+    }
   }
 
-  let requestCount = 0;
-
-  return async function convert(filter) {
-    const iframe = await createIframe();
-    const requestId = `request${requestCount++}`;
+  return result || { errors: ['Failed to initiate converter'], rules: [] };
+}
 
-    return new Promise((resolve) => {
-      requestResolvers.set(requestId, resolve);
+let offscreenTimeout = null;
+function closeOffscreenDocument() {
+  if (offscreenTimeout) {
+    clearTimeout(offscreenTimeout);
+  }
 
-      iframe.contentWindow.postMessage(
-        {
-          action: 'convert',
-          converter: 'adguard',
-          filters: [requestId, filter],
-        },
-        '*',
-      );
-    });
-  };
+  offscreenTimeout = setTimeout(() => {
```

Comment:
Just for understanding: why delay the close operation? Would it cause problems to do "chrome.offscreen.closeDocument()" immediately? If not, can there be races if the one second is not enough?